spi/ssd1309_spi/ssd1309_spi.c

```diff
@@ -0,0 +1,144 @@
+/**
+ * Copyright (c) 2025 mjcross
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+**/
+
+// A simple example to display text on an ssd1309-based OLED display with
+// an SPI interface. It should also work on a ssd1306 device (not tested).
+//
+// To understand the commands and addressing modes for the display refer to
+// the manufacturer's datasheet at https://www.hpinfotech.ro/SSD1309.pdf
+
+
+#include <stdio.h>
+#include "pico/stdlib.h"
+#include "hardware/spi.h"
+#include "ctype.h"
+#include "ssd1309_font.h"
+
+
+// We will use the spi0 peripheral and the following GPIO pins (see the
+// GPIO function select table in the RPi Pico datasheet).
+#define SPI_DEVICE      spi0
+#define PIN_CS          17      // chip select (active low)
+#define PIN_SCK         18      // SPI clock
+#define PIN_MOSI        19      // SPI data transmit (MOSI)
+#define PIN_DC          20      // data/command mode (low for command)
+#define PIN_R           21      // reset (active low)
+
+// ssd1309 accepts a maximum SPI clock rate of 10 Mbit/sec 
+#define SPI_BITRATE     10 * 1000 * 1000
+
+// dimensions of the display
+#define NUM_X_PIXELS    128
+#define NUM_Y_PIXELS    64
+
+// modes for the data/command pin
+#define DC_COMMAND_MODE 0
+#define DC_DATA_MODE    1
+
+
+// send a command byte to the display
+void send_command(uint8_t cmd_byte) {
+    gpio_put(PIN_DC, DC_COMMAND_MODE);
+    spi_write_blocking(SPI_DEVICE, &cmd_byte, 1);
+}
+
+// send a data byte to the display
+void send_data(uint8_t data_byte) {
+    gpio_put(PIN_DC, DC_DATA_MODE);
+    spi_write_blocking(SPI_DEVICE, &data_byte, 1);
+}
+
+// set the text cursor position (row, col)
```

Comment:
Is it worth clarifying what the maximum text row and column values are?

Reply:
fixed in https://github.com/raspberrypi/pico-examples/pull/731/commits/622210dc2773cd64dfdb7911f4b0633e6fb69551